packages/cli/test/snapshot-matcher.js

```diff
@@ -0,0 +1,205 @@
+// Copyright IBM Corp. 2019. All Rights Reserved.
+// Node module: @loopback/cli
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+'use strict';
+
+/*
+A lightweight helper for snapshot-based testing.
+- Feature inspired by https://jestjs.io/docs/en/snapshot-testing
+- Implementation inspired by node-tap
+
+This is an initial experimental version. In the (near) future, we should
+move this file to a standalone package so that all Mocha users can use it.
+*/
+
+const assert = require('assert');
+const path = require('path');
+
+module.exports = {
+  initializeSnapshots,
+};
+
+/**
+ * Create a function to match the given value against a pre-recorder snapshot.
+ *
+ * Example usage:
+ *
+ * ```js
+ * // At the top of your test file, initialize the matcher function.
+ * // You can also move this code to a shared file require()d from tests.
+ * const expectToMatchSnapshot = initializeSnapshots(
+ *   path.resolve(__dirname, '../../__snapshots__'),
+ * );
+ *
+ * // Inside your tests, call `expectToMatchSnapshot(actualValue)`
+ * describe('my feature', () => {
+ *   it('works for strings', function() {
+ *     expectToMatchSnapshot('foo\nbar');
+ *   });
+ * });
+ * ```
+ */
+function initializeSnapshots(snapshotDir) {
+  let currentTest;
+
+  beforeEach(function setupSnapshots() {
+    // eslint-disable-next-line no-invalid-this
+    currentTest = this.currentTest;
+    currentTest.__snapshotCounter = 1;
+  });
+
+  if (!process.env.UPDATE_SNAPSHOTS) {
+    return function expectToMatchSnapshot(actual) {
+      matchSnapshot(snapshotDir, currentTest, actual);
+    };
+  }
+
+  const snapshots = Object.create(null);
+  after(function updateSnapshots() {
+    for (const f in snapshots) {
+      const snapshotFile = buildSnapshotFilePath(snapshotDir, f);
+      writeSnapshotData(snapshotFile, snapshots[f]);
+    }
+  });
+
+  return function expectToRecordSnapshot(actual) {
+    recordSnapshot(snapshots, currentTest, actual);
+  };
+}
+
+function matchSnapshot(snapshotDir, currentTest, actualValue) {
+  assert(
+    typeof actualValue === 'string',
+    'Snapshot matcher supports string values only, but was called with ' +
+      typeof actualValue,
+  );
+
+  const snapshotFile = buildSnapshotFilePath(snapshotDir, currentTest.file);
+  const snapshotData = loadSnapshotData(snapshotFile);
+  const key = buildSnapshotKey(currentTest);
+
+  if (!(key in snapshotData)) {
+    throw new Error(
+      `No snapshot found for ${JSON.stringify(key)}.\n` +
+        'Run the tests with UPDATE_SNAPSHOTS=1 in the environment ' +
+        'to create and update snapshot files.',
+    );
+  }
+
+  assert.deepStrictEqual(actualValue, snapshotData[key]);
```

Comment:
FIXME: when this assertion fail, we should tell the user how to fix the snapshots via `UPDATE_SNAPSHOTS=1`.

Reply:
See https://github.com/strongloop/loopback-next/pull/3776